exotations/solvers/exotica_ddp_solver/include/exotica_ddp_solver/abstract_ddp_solver.h

```diff
@@ -0,0 +1,217 @@
+
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+
+    // initialize Gain matrices
+    K_gains_.assign(T, Eigen::MatrixXd(NU, N));
+    k_gains_.assign(T, Eigen::VectorXd(NU, 1));
+
+    // all of the below are not pointers, since we want to copy over
+    //  solutions across iterations
+    Eigen::MatrixXd new_U, global_best_U;
+    solution.resize(T, NU);
+
+    if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Running DDP solver for max " << GetNumberOfMaxIterations() << " iterations");
+
+    double last_cost = 0, global_best_cost = 0;
+    int last_best_iteration = 0;
+
+    for (int iteration = 0; iteration < GetNumberOfMaxIterations(); ++iteration)
+    {
+        // Backwards pass computes the gains
+        BackwardPass();
+        if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Backward pass complete");
+
+        // forward pass to compute new control trajectory
+        // TODO: Configure line-search space from xml
+        // TODO (Wolf): What you are doing is a forward line-search when we may try a
+        //    backtracking line search for a performance improvement later - this just as an aside.
+        const Eigen::VectorXd alpha_space = Eigen::VectorXd::LinSpaced(10, 0.1, 1.0);
+        double current_cost = 0, best_alpha = 0;
+
+        Eigen::MatrixXd ref_trajectory = prob_->get_X();
+        // perform a linear search to find the best rate
+        for (int ai = 0; ai < alpha_space.rows(); ++ai)
+        {
+            double alpha = alpha_space(ai);
+            double cost = ForwardPass(alpha, ref_trajectory);
+
+            if (ai == 0 || cost < current_cost)
+            {
+                current_cost = cost;
+                new_U = prob_->get_U();
+                best_alpha = alpha;
+            }
+        }
+
+        if (debug_)
+        {
+            HIGHLIGHT_NAMED("DDPSolver", "Forward pass complete with cost: " << current_cost << " and alpha " << best_alpha);
+            HIGHLIGHT_NAMED("DDPSolver", "Final state: " << prob_->get_X(T - 1).transpose());
+        }
+
+        // copy solutions for next iteration
+        if (iteration == 0 || global_best_cost > current_cost)
+        {
+            global_best_cost = current_cost;
+            last_best_iteration = iteration;
+            global_best_U = new_U;
+        }
+
+        if (iteration - last_best_iteration > base_parameters_.FunctionTolerancePatience)
+        {
+            if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Early stopping criterion reached.");
+            break;
+        }
+
+        if (last_cost - current_cost < base_parameters_.FunctionTolerance && last_cost - current_cost > 0)
+        {
+            if (debug_) HIGHLIGHT_NAMED("DDPSolver", "Function tolerance reached.");
+            break;
+        }
+
+        if (iteration == GetNumberOfMaxIterations() - 1 && debug_) HIGHLIGHT_NAMED("DDPSolver", "Max iterations reached.");
```

Comment:
Put `debug_` at the beginning of the condition for quicker lazy evaluation.

Reply:
We want to break regardless of whether `debug_` is true.

Reply:
`if (debug_ && iteration == GetNumberOfMaxIterations() - 1)`

Reply:
I was looking at the wrong thing..